File: CIVITASRUBY/lib/juego_texto/controlador.rb
#encoding:utf-8

module Civitas
  require_relative 'vista_textual'
  require_relative '../civitas_juego'
  require_relative '../operaciones_juego'
  require_relative '../operacion_inmobiliaria'
  require_relative '../salidas_carcel'
  class Controlador
    GESTIONES = [Gestiones_Inmobiliarias::VENDER,Gestiones_Inmobiliarias::HIPOTECAR,Gestiones_Inmobiliarias::CANCELAR_HIPOTECA,Gestiones_Inmobiliarias::CONSTRUIR_CASA,Gestiones_Inmobiliarias::CONSTRUIR_HOTEL,Gestiones_Inmobiliarias::TERMINAR]

    def initialize(juego, vista)
      @vista = vista
      @civitas = juego
    end
    
    def juega
      @vista.set_civitas_juego(@civitas)
      
      while @civitas.final_del_juego == false
        @vista.pausa
        op = @civitas.siguiente_paso
        @vista.mostrar_siguiente_operacion(op)
        
        if op != Operaciones_juego::PASAR_TURNO
          @vista.mostrar_eventos
        end
        
        if @civitas.final_del_juego == true
          jugadores = @civitas.ranking_UI
          
          @vista.mostrar_estado("Ranking \n")
          jugadores.each do |j|
            @vista.mostrar_estado(j)
          end
          
        else
          
          case op
          when Operaciones_juego::COMPRAR
            r = @vista.comprar
            if r == Respuestas::SI
              @civitas.comprar
              
            end
            @civitas.siguiente_paso_completado(op)
          when Operaciones_juego::GESTIONAR
            @vista.gestionar
            i_gestion = @vista.get_gestion
            i_propiedad = @vista.get_propiedad
            operacion = Operacion_Inmobiliaria.new(GESTIONES[i_gestion],i_propiedad)
            case operacion.gestion
            when GESTIONES[0]
              @civitas.vender(i_propiedad)
              
            when GESTIONES[1]
              @civitas.hipotecar(i_propiedad)
            when GESTIONES[2]
              @civitas.cancelar_hipoteca(i_propiedad)
            when GESTIONES[3]
              @civitas.construir_casa(i_propiedad)
            when GESTIONES[4]
              @civitas.construir_hotel(i_propiedad)
            when GESTIONES[5]
              @civitas.siguiente_paso_completado(op)
            end
           
          when Operaciones_juego::SALIR_CARCEL
            if @vista.salir_carcel == Salidas_carcel::PAGANDO
              @civitas.salir_carcel_pagando
            else
              @civitas.salir_carcel_tirando
            end
             @civitas.siguiente_paso_completado(op)
          end
        end
        
      end
    end
  end
end

File: CIVITAS/src/JuegoTexto/Controlador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package JuegoTexto;

import civitas.CivitasJuego;
import civitas.OperacionesJuego;
import java.util.ArrayList;
import civitas.Jugador;
import civitas.Respuestas;
import civitas.OperacionInmobiliaria;
import civitas.GestionesInmobiliarias;
import civitas.SalidasCarcel;
/**
 *
 * @author Usuario
 */

public class Controlador {
    private VistaTextual vista;
    private CivitasJuego civitas;

    Controlador( CivitasJuego juego , VistaTextual vista) {
        this.vista = vista;
        this.civitas = juego;
    }
    
    void juega(){
        this.vista.setCivitasJuego(civitas);
        
        while(!this.civitas.finalDelJuego()){
            this.vista.pausa();
            OperacionesJuego op = this.civitas.siguientePaso();
            
            
            this.vista.mostrarSiguienteOperacion(op);
            
            if(!op.equals(OperacionesJuego.PASAR_TURNO)){
                this.vista.mostrarEventos();
            }
            
            if(this.civitas.finalDelJuego()){
                ArrayList<String> jugadores = this.civitas.ranking_UI();
                this.vista.mostrarEstado("Ranking: \n");
                for(String j : jugadores){
                    this.vista.mostrarEstado(j);
                }
             
            }else{
                switch(op){
                    case COMPRAR:
                        Respuestas r = this.vista.comprar();
                        if(r.equals(Respuestas.SI))
                            this.civitas.comprar();
                        
                        this.civitas.siguientePasoCompletado(op);
                    
                        
                        break;
                        
                    case GESTIONAR:
                      
                        this.vista.gestionar();
                        int igestion = this.vista.getGestion();
                        int ipropiedad = this.vista.getPropiedad();
                        OperacionInmobiliaria operacion = new OperacionInmobiliaria(GestionesInmobiliarias.values()[igestion],ipropiedad);
                        switch(operacion.getGestion()){
                            case VENDER:
                                this.civitas.vender(ipropiedad);
                                break;
                            case HIPOTECAR:
                                this.civitas.hipotecar(ipropiedad);
                                break;
                            case CANCELAR_HIPOTECA:
                                this.civitas.cancelarHipoteca(ipropiedad);
                                break;
                            case CONSTRUIR_CASA:
                                this.civitas.construirCasa(ipropiedad);
                                break;
                            case CONSTRUIR_HOTEL:
                                this.civitas.construirHotel(ipropiedad);
                                break;
                            case TERMINAR:
                                this.civitas.siguientePasoCompletado(op);
                                break;
                        }
                       
                        break;
                    case SALIR_CARCEL:
                        if(this.vista.salirCarcel().equals(SalidasCarcel.PAGANDO)){
                            this.civitas.salirCarcelPagando();
                        }else{
                            this.civitas.salirCarcelTirando();
                        }
                        this.civitas.siguientePasoCompletado(op);
                        break;
                        
                }
            }
        }
    }
    
    
    
}
